### PR TITLE
10684-formation-breadcrumb

### PR DESCRIPTION
**Pull request by 1Copenut**
## Purpose of the PR

This PR is asking to move all React apps in production to use the new Formation `Breadcrumb` component, instead of one-off components. It seeks to move the breadcrumb array logic into each app's parent container, and call it from a `renderBreadcrumbs()` method, with optional args passed in as needed.

The `renderBreadcrumbs()` method makes use of a `crumbs` array, populated with hard-coded `<a>` tags for static pages, and React Router `<Link>` components, to create a single outputted list of breadcrumb links that is easy to test in manual and automated settings.

## Added

* `<Breadcrumb>` component from Formation 1.2.0
* `renderBreadcrumb()` method to several React containers. Always returns `crumbs` array.
* e2e tests for breadcrumb functionality
* core-js lib to devDependencies to remove unit test error of missing ES7 pad-start method

## Updated

* 2 unit tests to pass in mocked `location` object
* All instances of app containers using breadcrumb components
* All breadcrumbs maintain the existing local breadcrumb components, and use the Formation Breadcrumb inside them for formatting and styles.
* Carefully added routes and updated text for app breadcrumb "You are here" pages.

## Removed

* Overloaded logic in the Claims and Appeals app for breadcrumb logic. This has been centralized in the app-specific `<NS>Breadcrumb` component.

## Screenshots

The following screen shots have been vetted with @mnorthuis and @bethpotts for content accuracy. Tagging them on the PR as well.

---

active-rx

---

disability-claims

---

gi-bill-comparison

### CODE REVIEW

File: src/applications/claims-status/containers/AdditionalEvidencePage.jsx [hunk 1]
```diff
@@ -54,15 +55,30 @@ class AdditionalEvidencePage extends React.Component {
...
+
+  renderBreadcrumbs(claim) {
+    const crumbs = [
+      <a href="/" key="home">Home</a>,
+      <a href="/disability-benefits/" key="disability-benefits">Disability Benefits</a>,
+      <Link to="/" key="claims-home">Track Your Claims and Appeals</Link>,
+      <Link to={`/your-claims/${claim.id}`} key="claim-id">Status Details</Link>,
+      <Link to={`/your-claims/${claim.id}/additional-evidence`} key="additional-evidence">Additional Evidence</Link>
+    ];
+
```
Comment (ncksllvn) [inline src/applications/claims-status/containers/AdditionalEvidencePage.jsx]: Is the `key` attribute of each element still necessary here?

File: src/applications/claims-status/containers/ClaimEstimationPage.jsx [hunk 2]
```diff
@@ -2,24 +2,37 @@ import React from 'react';
...
+
+    return crumbs;
+  }
+
+
   render() {
-    const claimType = !this.props.loading ? getClaimType(this.props.claim) : '';
+    // const claimType = !this.props.loading ? getClaimType(this.props.claim) : '';
+    const claim = this.props.claim;
     return (
```
Comment (ncksllvn) [inline src/applications/claims-status/containers/ClaimEstimationPage.jsx]: Is `claimType` not necessary here anymore? Asking because I see the check for `this.props.loading` removed. Have you tested to see what happens when you navigate to this page directly though the address? That might be where the `loading` prop comes in to play.

File: src/applications/claims-status/containers/DocumentRequestPage.jsx [hunk 3]
```diff
@@ -71,24 +70,36 @@ class DocumentRequestPage extends React.Component {
...
+      <a href="/" key="home">Home</a>,
+      <a href="/disability-benefits/" key="disability-benefits">Disability Benefits</a>,
+      <Link to="/" key="claims-home">Track Your Claims and Appeals</Link>,
+      <Link to={`/your-claims/${claim.id}`} key="claim-id">Status Details</Link>,
+      <Link to={`/your-claims/${claim.id}/document-request/${trackedItem}`} key="claim-id">Document Request</Link>
+    ];
+
+    return crumbs;
+  }
+
```
Comment (ncksllvn) [inline src/applications/claims-status/containers/DocumentRequestPage.jsx]: This seems pretty different than the previous breadcrumbs. Were there new requirements here? Just confirming that it's no longer necessary to link to `/closed` anymore based on `this.props.claim.attributes.open`. It also looks like the title of `claims-home` changed from `Your Claims` to `Track Your Claims and Appeals`.

File: src/applications/claims-status/containers/AskVAPage.jsx [hunk 4]
```diff
@@ -15,24 +15,42 @@ class AskVAPage extends React.Component {
...
+    const crumbs = [
+      <a href="/" key="home">Home</a>,
+      <a href="/disability-benefits/" key="disability-benefits">Disability Benefits</a>,
+      <Link to="your-claims" key="claims-home">Track Your Claims and Appeals</Link>,
+      <Link to={`your-claims/${claim.id}`} key="claim-id">Status Details</Link>,
+      <Link to={`your-claims/${claim.id}/ask-va-to-decide`} key="ask-va">Ask for Claim Decision</Link>
+    ];
+
+    return crumbs;
+  }
```
Comment (ncksllvn) [inline src/applications/claims-status/containers/AskVAPage.jsx]: I'm not sure if it's worth the trouble or not, but if this basic set of links is shared by most components/containers within this application, it might be worth keeping the application's breadcrumb component, and implementing the Formation Breadcrumbs within that. That way the breadcrumbs of this application remain centralized and if a link changes it only needs to change in a single location.

File: src/applications/claims-status/containers/YourClaimsPage.jsx [hunk 5]
```diff
@@ -200,7 +209,10 @@ class YourClaimsPage extends React.Component {
 
     return (
       <div className="claims-container">
-        <Breadcrumbs/>
+        <Breadcrumbs>
+          {this.renderBreadcrumbs()}
+          {this.props.children}
+        </Breadcrumbs>
         <div className="row">
```
Comment (ncksllvn) [inline src/applications/claims-status/containers/YourClaimsPage.jsx]: How come we have to pass `this.props.children`?

File: src/applications/facility-locator/tests/00-breadcrumbs.e2e.spec.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    client.waitForElementVisible('.va-nav-breadcrumbs-list', Timeouts.normal);
+    client.expect.element('.va-nav-breadcrumbs-list li:not(:nth-last-child(2))').to.have.css('display').which.equal('none');
+    client.expect.element('.va-nav-breadcrumbs-list li:nth-last-child(2)').to.have.css('display').which.equal('inline-block');
+
+    // Reset default size
+    client.resizeWindow(1024, 768);
+
+    client.end();
+  });
```
Comment (ncksllvn) [inline src/applications/facility-locator/tests/00-breadcrumbs.e2e.spec.js]: These E2E tests are great, but I'm not sure how necessary it is to test each application's implementation since they share a common component now.

File: src/applications/health-records/containers/HealthRecordsApp.jsx [hunk 7]
```diff
@@ -23,6 +25,26 @@ const AppContent = ({ children }) => (
...
+        <Link to="/" key="main">Get Your VA Health Records</Link>,
+        <Link to="/download/" key="download">Download Your Health Records</Link>
+      );
+    } else {
+      crumbs.push(<Link to="/" key="main">Get Your VA Health Records</Link>);
+    }
+
+    return crumbs;
+  }
+
```
Comment (ncksllvn) [inline src/applications/health-records/containers/HealthRecordsApp.jsx]: Was this logic extracted from the application's breadcrumbs?

**Review by ncksllvn — COMMENTED** [review]
Awesome work, love seeing this component come together. My concern is that each application may still need a method of centralizing their breadcrumbs (in their own breadcrumbs), even if behind-the-scenes each breadcrumb component is just proxying the work to the Formation Breadcrumb component. But then again I'm an outsider to many of these applications so that could be unnecessary. If so, it might be worth going ahead and deleting the unused breadcrumb components in this PR.

File: src/applications/claims-status/containers/AskVAPage.jsx (hunk 4 above)
Comment (1Copenut) [reply src/applications/claims-status/containers/AskVAPage.jsx]: @ncksllvn This is a great call, and I hadn't considered plumbing the Formation breadcrumb through the app's breadcrumb. I will refactor that, and look to retain the logic as it is now.

File: src/applications/claims-status/containers/DocumentRequestPage.jsx (hunk 3 above)
Comment (1Copenut) [reply src/applications/claims-status/containers/DocumentRequestPage.jsx]: @ncksllvn The name change was coordinated with the Content team. Removing `/closed` was an oversight, and based on the wider conversation to consolidate the app's breadcrumb using Formation, will be restored.

File: src/applications/claims-status/components/ClaimDetailLayout.jsx [hunk 8]
```diff
@@ -23,13 +34,12 @@ export default class ClaimDetailLayout extends React.Component {
 
     let content;
     if (!loading) {
-      const claimsPath = `your-claims${claim.attributes.open ? '' : '/closed'}`;
```
Comment (cehsu) [inline src/applications/claims-status/components/ClaimDetailLayout.jsx]: Is this change intentional?

File: src/applications/claims-status/containers/AdditionalEvidencePage.jsx [hunk 9]
```diff
@@ -71,16 +87,15 @@ class AdditionalEvidencePage extends React.Component {
     } else {
       const claim = this.props.claim;
       const filesPath = `your-claims/${claim.id}/files`;
-      const claimsPath = `your-claims${claim.attributes.open ? '' : '/closed'}`;
+      // const claimsPath = `your-claims${claim.attributes.open ? '' : '/closed'}`;
```
Comment (cehsu) [inline src/applications/claims-status/containers/AdditionalEvidencePage.jsx]: Similar to above, is this change intentional? If so, should the comment be removed?

File: src/applications/claims-status/containers/AskVAPage.jsx (hunk 4 above)
Comment (cehsu) [reply src/applications/claims-status/containers/AskVAPage.jsx]: Consolidating shared logic is a good idea. It looks like the links differ slightly, perhaps the shared component could accept props to determine which optional links to include, in order to accommodate this.

File: src/applications/facility-locator/tests/00-breadcrumbs.e2e.spec.js (hunk 6 above)
Comment (cehsu) [reply src/applications/facility-locator/tests/00-breadcrumbs.e2e.spec.js]: The form [e2e tests](https://github.com/department-of-veterans-affairs/vets-website/blob/18157605a659fae278754771efba6ab050eb60b8/src/applications/pensions/tests/00-all-fields.e2e.spec.js#L26) include checks to ensure that the progress bar presents the appropriate information, it seems consistent to use e2e tests to ensure that the breadcrumbs display the appropriate information as well.

File: src/applications/health-records/sass/health-records.scss [hunk 10]
```diff
@@ -22,9 +22,9 @@
     margin: 1em 0 !important;
   }
 
-  a:visited {
-    color: $color-primary;
-  }
+  // a:visited {
```
Comment (cehsu) [inline src/applications/health-records/sass/health-records.scss]: Should this comment be removed?

File: src/applications/health-records/containers/HealthRecordsApp.jsx [hunk 11]
```diff
@@ -1,14 +1,16 @@
...
 
 import DowntimeNotification, { externalServices } from '../../../platform/monitoring/DowntimeNotification';
+import Breadcrumbs from '@department-of-veterans-affairs/formation/Breadcrumbs';
 import Modal from '@department-of-veterans-affairs/formation/Modal';
 import MHVApp from '../../../platform/user/authorization/containers/MHVApp';
 import backendServices from '../../../platform/user/profile/constants/backendServices';
 import RequiredLoginView from '../../../platform/user/authorization/components/RequiredLoginView';
 import { closeModal } from '../actions/modal';
-import Breadcrumbs from '../components/Breadcrumbs';
+// import Breadcrumbs from '../components/Breadcrumbs';
```
Comment (cehsu) [inline src/applications/health-records/containers/HealthRecordsApp.jsx]: Similar to above, could you please remove the existing breadcrumb component so that we can review the diff and check how this logic relates to the logic it is replacing?

File: src/applications/rx/containers/RxRefillsApp.jsx [hunk 12]
```diff
@@ -1,16 +1,18 @@
 import PropTypes from 'prop-types';
 import React from 'react';
+import _ from 'lodash';
 import { connect } from 'react-redux';
+import { Link } from 'react-router';
 
 import backendServices from '../../../platform/user/profile/constants/backendServices';
 import DowntimeNotification, { externalServices } from '../../../platform/monitoring/DowntimeNotification';
 import MHVApp from '../../../platform/user/authorization/containers/MHVApp';
-import Breadcrumbs from '../components/Breadcrumbs';
```
Comment (cehsu) [inline src/applications/rx/containers/RxRefillsApp.jsx]: Similar to above, regarding the visibility of the diff.

**Review by cehsu — COMMENTED** [review]
This looks great, I had a few questions/ideas based off of the review conversation so far.

**Comment by ncksllvn** [comment]
> core-js lib to devDependencies to remove unit test error of missing ES7 pad-start method

@1Copenut Can you expand on this? @bshyong and are seeing the error below while trying to [upgrade Formation to `1.2.2`](https://github.com/department-of-veterans-affairs/vets-website/pull/8057).

```
Error: Cannot find module 'core-js/modules/es7.string.pad-start'

    at Function.Module._resolveFilename (module.js:547:15)
```

File: src/applications/claims-status/containers/ClaimEstimationPage.jsx (hunk 2 above)
Comment (1Copenut) [reply src/applications/claims-status/containers/ClaimEstimationPage.jsx]: @ncksllvn I loaded this page directly a few times, and the claimType was just updating the breadcrumb link name. Since we've simplified it to `Status Details` on all claim and appeals detail views, I feel okay removing it.

File: src/applications/claims-status/components/ClaimDetailLayout.jsx (hunk 8 above)
Comment (1Copenut) [reply src/applications/claims-status/components/ClaimDetailLayout.jsx]: I've restored the previous logic, but I'm not sure we need it. The `routes.jsx` file only had routing for `/closed` in commented out, old versions of the file. Doesn't hurt to leave it though.

File: src/applications/facility-locator/tests/00-breadcrumbs.e2e.spec.js (hunk 6 above)
Comment (1Copenut) [reply src/applications/facility-locator/tests/00-breadcrumbs.e2e.spec.js]: @cehsu Great call! I updated the e2e tests to check not only for presence of the breadcrumbs, but the text of the terminal element on standard and mobile breadcrumbs. Makes the tests more comprehensive.

File: src/applications/gi/components/heading/Breadcrumbs.jsx [hunk 13]
```diff
@@ -1,42 +1,56 @@
...
-      crumbs.push(<Link to={root} key="main">GI Bill® Comparison Tool</Link>);
+    if (pathname.match(/search/)) {
+      crumbs.push(<Link to={`/search/${search}`} key="search-results">Search Results</Link>);
     }
 
     if (pathname.match(/profile/)) {
       if (this.props.includeSearch) {
-        crumbs.push(<a onClick={browserHistory.goBack} key="search-results">Search Results</a>);
+        crumbs.push(
+          <a
```
Comment (cehsu) [inline src/applications/gi/components/heading/Breadcrumbs.jsx]: Could we make this anchor element one line?

File: src/applications/gi/components/heading/Breadcrumbs.jsx [hunk 14]
```diff
@@ -1,42 +1,56 @@
...
+            onClick={browserHistory.goBack}
+            key="search-results">Search Results
+          </a>,
+          <Link to={`/profile/${facilityCode}`} key="result-detail">School Details</Link>
+        );
+      } else {
+        // User copy-pastes a profile URL, or refreshes page
+        crumbs.push(
```
Comment (cehsu) [inline src/applications/gi/components/heading/Breadcrumbs.jsx]: Could we move this `crumbs.push` statement beneath `if (pathname.match(/profile/)) {` so we can avoid duplicating it?

File: src/applications/rx/containers/RxRefillsApp.jsx [hunk 15]
```diff
@@ -5,7 +5,7 @@ import { connect } from 'react-redux';
...
+import RxBreadcrumbs from '../components/Breadcrumbs';
```
Comment (cehsu) [inline src/applications/rx/containers/RxRefillsApp.jsx]: If we are planning to name the breadcrumb components in each app, should we consider using the same name as the 'App' component, such as RxRefillsApp (so RxRefillsBreadcrumbs)?

File: src/applications/rx/tests/components/Breadcrumbs.unit.spec.jsx [hunk 16]
```diff
@@ -17,31 +17,54 @@ const props = {
   }
 };
 
+const crumbs = [
+  <a href="/" key="home">Home</a>,
+  <a href="/health-care/" key="healthcare">Health Care</a>,
+  <Link to="/" key="prescriptions">Prescription Refills</Link>
+];
+
 describe('<Breadcrumbs>', () => {
```
Comment (cehsu) [inline src/applications/rx/tests/components/Breadcrumbs.unit.spec.jsx]: The test names typically reflect the component names, if we update the component names, we will probably want to update these as well.

File: src/applications/rx/components/Breadcrumbs.jsx [hunk 17]
```diff
@@ -25,50 +26,28 @@ class Breadcrumbs extends React.Component {
     const crumbs = [
       <a href="/" key="home" onClick={() => { recordEvent({ event: 'nav-breadcrumb' });}}>Home</a>,
       <a href="/health-care/" key="healthcare">Health Care</a>,
+      <Link to="/" key="prescriptions">Prescription Refills</Link>
     ];
 
-    if (pathname.match(/\/\d+$/)) {
-      crumbs.push(<Link to="/" key="prescriptions">Prescription Refills</Link>);
+    if (pathname.match(/\/\d+\/?(track)?$/)) {
```
Comment (cehsu) [inline src/applications/rx/components/Breadcrumbs.jsx]: @1Copenut could you please comment on these differences from the earlier version?

File: src/applications/health-records/components/Breadcrumbs.jsx [hunk 18]
```diff
@@ -1,26 +1,26 @@
...
       <a href="/health-care/" key="healthcare">Health Care</a>,
+      <Link to="/" key="main">Get Your VA Health Records</Link>
     ];
 
     if (pathname.match(/download\/?$/)) {
-      crumbs.push(<Link to="/" key="main">Get Your VA Health Records</Link>);
-      crumbs.push(<span key="download"><strong>Download Your Health Records</strong></span>);
-    } else {
-      crumbs.push(<span key="main"><strong>Get Your VA Health Records</strong></span>);
+      crumbs.push(<Link to="/download" key="download">Download Your Health Records</Link>);
```
Comment (cehsu) [inline src/applications/health-records/components/Breadcrumbs.jsx]: @1Copenut could you please comment on these differences?

**Review by cehsu — COMMENTED** [review]
Thanks for sharing these updates @1Copenut, I had a few questions/comments for your consideration.

**Comment by ncksllvn** [comment]
Closing, because I think the approach for this changed and was implemented here instead - https://github.com/department-of-veterans-affairs/vets-website/pull/7976. Apologies if this PR was finished @1Copenut, feel free to reopen! Just trying to get cleaned up.
